### PR TITLE
Fix "Uncaught ReferenceError: ext is not defined"

### PR DESCRIPTION
**Pull request by Rob--W**
This patch is temporary and can be reverted once the Chromium bug
(https://crbug.com/416907) has been fixed.

To make merging/reverting the patch easier, I have changed the main if-block
to a function that takes the original if-body as a main function. The main
function is called whenever the bootstrapper has made sure that the environment
is ready. (The alternatives, disabling the content script or using XHR, were
not so appealing.)

Fixes
https://issues.adblockplus.org/ticket/1370
https://adblockplus.org/forum/viewtopic.php?f=10&t=25507

I hereby waive any rights of this contributed piece of software, you may do whatever you want with it (this comment is an alternative to signing the CLA at https://adblockplus.org/eyeo-contributor-license-agreement.pdf).

### CODE REVIEW

**Comment by snoack** [comment]
Hi Rob, thanks for taking over the Chromium bug, I reported. Given that you submitted a patch with a workaround for Adblock Plus, I assume that it won't be fixed upstream any time soon? :-/

Considering the complexity of the workaround, and given this is a minor bug for Adblock Plus with no user visible effect (beside a JS error in the console) reported so far, I would prefer to just ignore the issue in Adblock Plus for now.

**Comment by Rob--W** [comment]
Unless my Chromiums patch gets merged back to beta (or even stable), the bug will be around for another 2 months.

I'd prefer if you merge the patch, because "the JS error in the console" _is_ quite a nuisance for web developers. And the issue may be user-visible as well, because Adblock Plus skips iframe checking in Chrome 37+.

If you discount the elaborate comments and line breaks, the patch is not even 20 lines. Reviewing that is doable, right? The patch just fetches the content script data, evaluates the code, then executes the main function.

If you don't like the complexity, I could also submit a new PR that tests whether `ext` exists before trying to continue with the function. Then the error will also go away, but the ad detection will be slightly inferior (compared to Chrome 36-).

**Comment by snoack** [comment]
So the patch fixing the Chromium bug is on the way? (There is no review link in the Chromium issue)

It's still a corner case, which only occurs on pages that use frames with `javascript:` URLs that evaluate to a string. And even if we add a workaround, we might not release another version before Chrome 39 is released. And you sound like it will be fixed at latest with Chrome 40 anyway.

**Comment by palant** [comment]
Frankly, I don't like the added complexity here either (I'm actually unhappy about the existing complexity in this area already). The functionality lost due to this issue is something that didn't even exist until recently and users pretty much never noticed. So I agree with Sebastian - better wait for this to be fixed in Chrome. If anything, we should merely get rid of the error message and wait for the underlying issue to be fixed in Chrome (two months is much less than what we typically have to wait).

**Comment by Rob--W** [comment]
I've replaced the fix (https://github.com/Rob--W/adblockpluschrome/commit/02bd51c782d36488d71faa7082f68e211cbb59a2) with a small patch that merely avoids the error message (https://github.com/Rob--W/adblockpluschrome/commit/eb73729ab150539e6d436ed1f580a7f9211b5132).

@wallunit The patch is here: https://codereview.chromium.org/684143002/.

File: include.postload.js [hunk 1]
```diff
@@ -472,7 +472,10 @@ function removeDotSegments(u) {
   }
 }
 
-if (document instanceof HTMLDocument)
+// In Chrome 37-39, the document_end content script (this one) runs properly, while the 
+// document_start content scripts (that defines ext) do not. Check whether variable ext
+// exists before continuing to avoid "Uncaught ReferenceError: ext is not defined".
+if (typeof ext == 'object' && document instanceof HTMLDocument)
```
Comment (snoack) [inline include.postload.js]: Our coding style guidelines require double quotes, moreover I'd prefer to check `"ext" in window` instead.

Comment (Rob--W) [inline include.postload.js]: Done.

**Comment by snoack** [comment]
LGTM, from my side.

**Comment by palant** [comment]
This has been pushed:

​https://hg.adblockplus.org/adblockpluschrome/rev/e60ae9f70fd2
​https://hg.adblockplus.org/adblockpluschrome/rev/44ea44f6c195

The additional commit is me improving the comment slightly: it needs to reference the Chrome issue so that it's easier to find, and the original wording implied that `document_start` scripts would _never_ run which isn't true.

These changes will make their way to GitHub shortly.

**Comment by Rob--W** [comment]
The patch (https://codereview.chromium.org/684143002/) missed the boat by a few days, it will be fixed in Chrome 41 (not Chrome 40).

**Comment by palant** [comment]
Updated comment to reflect that: https://hg.adblockplus.org/adblockpluschrome/rev/9430b640fc04

**Pull request closed without merging**